src/main/java/org/opensearch/neuralsearch/processor/NeuralSparseTwoPhaseProcessor.java

```diff
@@ -0,0 +1,223 @@
+import org.opensearch.search.rescore.QueryRescorerBuilder;
+import org.opensearch.search.rescore.RescorerBuilder;
+
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * A SearchRequestProcessor to generate two-phase NeuralSparseQueryBuilder,
+ * and add it to the Rescore of a searchRequest.
+ */
+@Setter
+@Getter
+public class NeuralSparseTwoPhaseProcessor extends AbstractProcessor implements SearchRequestProcessor {
+
+    public static final String TYPE = "neural_sparse_two_phase_processor";
+    private boolean enabled;
+    private float ratio;
+    private float window_expansion;
+    private int max_window_size;
+    private static final String PARAMETER_KEY = "two_phase_parameter";
+    private static final String RATIO_KEY = "prune_ratio";
+    private static final String ENABLE_KEY = "enabled";
+    private static final String EXPANSION_KEY = "expansion_rate";
+    private static final String MAX_WINDOW_SIZE_KEY = "max_window_size";
+
+    protected NeuralSparseTwoPhaseProcessor(
+        String tag,
+        String description,
+        boolean ignoreFailure,
+        boolean enabled,
+        float ratio,
+        float window_expansion,
+        int max_window_size
+    ) {
+        super(tag, description, ignoreFailure);
+        this.enabled = enabled;
+        if (ratio < 0f || ratio > 1f) {
+            throw new IllegalArgumentException(
+                String.format(Locale.ROOT, "The two_phase_parameter.prune_ratio must be within [0, 1]. Received: %f", ratio)
+            );
+        }
+        this.ratio = ratio;
+        if (window_expansion < 1.0f) {
+            throw new IllegalArgumentException(
+                String.format(Locale.ROOT, "The two_phase_parameter.expansion_rate must >= 1.0. Received: %f", window_expansion)
+            );
+        }
+        this.window_expansion = window_expansion;
+        if (max_window_size < 50) {
+            throw new IllegalArgumentException(
+                String.format(Locale.ROOT, "The two_phase_parameter.max_window_size must >= 50. Received: %n" + max_window_size)
+            );
+        }
+        this.max_window_size = max_window_size;
+    }
+
+    /**
+     * Process the search request of neural_sparse_two_phase_processor
+     * @param request the search request (which may have been modified by an earlier processor)
+     * @return request the search request that add the two-phase rescore query of neural sparse query.
+     */
+    @Override
+    public SearchRequest processRequest(final SearchRequest request) throws Exception {
+        if (!enabled || ratio == 0f) {
+            return request;
+        }
+        QueryBuilder queryBuilder = request.source().query();
+        // Collect the nested NeuralSparseQueryBuilder in the whole query.
+        Multimap<NeuralSparseQueryBuilder, Float> queryBuilderMap = ArrayListMultimap.create();
```

Comment:
This line is redundant since in the `collectNeuralSparseQueryBuilder` there'll be a map been created in the beginning.

Reply:
Removed.